actionview/lib/action_view/helpers/form_tag_helper.rb

```diff
@@ -106,15 +106,16 @@ def field_id(object_name, method_name, *suffixes, index: nil)
         sanitized_method_name = method_name.to_s.delete_suffix("?")
 
         # a little duplication to construct fewer strings
-        if sanitized_object_name.empty?
-          sanitized_method_name
-        elsif suffixes.any?
-          [sanitized_object_name, index, sanitized_method_name, *suffixes].compact.join("_")
-        elsif index
-          "#{sanitized_object_name}_#{index}_#{sanitized_method_name}"
-        else
-          "#{sanitized_object_name}_#{sanitized_method_name}"
-        end
+        id =
+          if sanitized_object_name.empty?
+            sanitized_method_name
+          elsif index
+            "#{sanitized_object_name}_#{index}_#{sanitized_method_name}"
+          else
+            "#{sanitized_object_name}_#{sanitized_method_name}"
+          end
+
+        [namespace, id, *suffixes].tap(&:compact!).join("_")
```

Comment:
Since `namespace` serves as a prefix, and `suffixes` serve as a suffix, this felt like the best way to sandwich the id between the two.

Since this block mentions `# a little duplication to construct fewer strings`, I had considered building up the `id` value with a `StringIO` instance and some conditionals, but I wasn't sure what would be most inline with the current allocation characteristics.

Would allocating an Array, splatting in var-args, self compacting and joining be that much worse than in-place String interpolation? Would the allocation benefits of shoveling onto a `StringIO` instance offset the code maintenance burden?

Reply:
Speaking of allocations, would it be more efficient to change the `delete_suffix` and `gsub` calls to their `!` alternatives?

```diff
diff --git a/actionview/lib/action_view/helpers/form_tag_helper.rb b/actionview/lib/action_view/helpers/form_tag_helper.rb
index 080816c9bc..e643c764ff 100644
--- a/actionview/lib/action_view/helpers/form_tag_helper.rb
+++ b/actionview/lib/action_view/helpers/form_tag_helper.rb
@@ -101,21 +101,24 @@ def field_id(object_name, method_name, *suffixes, index: nil, namespace: nil)
           object_name = object_name.model_name.singular
         end
 
-        sanitized_object_name = object_name.to_s.gsub(/\]\[|[^-a-zA-Z0-9:.]/, "_").delete_suffix("_")
-
-        sanitized_method_name = method_name.to_s.delete_suffix("?")
-
         # a little duplication to construct fewer strings
-        id =
-          if sanitized_object_name.empty?
-            sanitized_method_name
-          elsif index
-            "#{sanitized_object_name}_#{index}_#{sanitized_method_name}"
-          else
-            "#{sanitized_object_name}_#{sanitized_method_name}"
-          end
-
-        [namespace, id, *suffixes].tap(&:compact!).join("_")
+        sanitized_object_name = object_name.to_s.dup
+        sanitized_object_name.gsub!(/\]\[|[^-a-zA-Z0-9:.]/, "_")
+        sanitized_object_name.delete_suffix!("_")
+
+        sanitized_method_name = method_name.to_s.dup
+        sanitized_method_name.delete_suffix!("?")
+
+        parts = [
+          namespace,
+          sanitized_object_name.presence,
+          (index unless sanitized_object_name.empty?),
+          sanitized_method_name,
+          *suffixes
+        ]
+        parts.compact!
+
+        parts.join("_")
       end
```

To account for frozen string literals, there are `.dup` calls chained off the `to_s`. Maybe that offsets and allocation gains from the `!` variations.

Reply:
> Maybe that offsets and allocation gains from the `!` variations.

Yes, it would, unfortunately.  The `delete_suffix!` on `sanitized_object_name` *would* save an allocation after the string has already been `gsub`'d or `dup`'d, but that's probably not worth the extra code complexity.